cleverhans/model_abs.py

```diff
@@ -0,0 +1,73 @@
+from abc import ABCMeta
+from keras.models import Model
+
+
+class ModelAbstraction:
+    """
+    An abstract interface for a model that exposes the required functionalities
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
+        """
+        :param model: A function that takes a symbolic input and returns the
+                      symbolic output for the model's predictions.
+        """
+        raise NotImplementedError('Model abstraction not implementated.')
+
+    def get_hidden(self, layer):
```

Comment:
I see. Makes sense to pass `x`. The PR is actually returning a new Model for Keras rather than a symbolic representation which is not good.

But as for not passing `layer`, there would be an overhead of creating too many new Model classes to expose the features. Also, as for in-place operations, this would increase the computation time and memory usage.

How about `def fprop(self, x, layer)`?